keps/sig-instrumentation/3466-kubernetes-component-health-slis/README.md

```diff
@@ -0,0 +1,313 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP intends to allow us to emit health check data in a structured and consistent way,
+so that monitoring agents can consume healthz/livez/readyz data and create SLOs and alerts
+off of these inputs.
+
+## Motivation
+
+Healthchecking data is currently surfaced in unstructured format and is scraped by monitoring
+agents (as well as the kubelet), which must be configured to interpret the health data and
+to act upon it. This process does not lend itself readily to the creation of availability SLOs,
+since we basically require an outside agent to parse the health endpoint and convert this signal
+into an SLI.
+
+Instead, we propose to emit health check data in a structured format so that it can be ingested
+directly as availability SLIs, since that is what our health endpoints represent, i.e. the 
+health of a given Kubernetes component.
+
+### Goals
+
+- Create a uniform interface by which we can consume health checking information
+- Allow availability SLIs to be created without a specialized monitoring agent
+
+### Non-Goals
+
+- Creation of SLOs are out of scope. This KEP specifically targets the creation of signals which can be used as SLIs.
+
+## Proposal
+
+We are proposing to add a new endpoint on Kubernetes components `/metrics/health` which returns
+health check data in prometheus format.
+
+
+### Risks and Mitigations
+
+This is a separate endpoint, it does not have to be used. The risk of adding metrics is generally 
+cardinality, but in this case we are proposing known dimensions to the metrics, specifically:
+
+1. status - one of `Success`, `Error`, `Pending`
+1. type - one of `livez`, `readyz`, `healthz`
+1. name - the known name of the health check. AFAIK, these are all static strings in the Kubernetes codebase, therefore bounded in cardinality
+
+
+## Design Details
```

Comment:
I don't understand this design. Can you give an example call of this endpoint and what it would return?

Is the point of this to make health information have some history so it needn't be scraped as often?

If you are making a new metric, why wouldn't it go with other metrics (on the same page)?

My first thought is a pair of metrics, "healthy seconds" and "unhealthy seconds", that are both counters that count up. Was this considered?

Reply:
I will amend this with the expected output of the endpoint.

No, the point of this is to add a metrics endpoint so that we can generate component healthiness SLIs and scrape it **more often** than the normal metrics endpoint. The fact that it is scraped more frequently is a result of the fact that it is an SLI, and SLIs tend to require greater granularity of data points and also the fact that the cardinality of these metrics are small, so scraping at a higher frequency shouldn't increase storage costs that much. If we were to add these metrics to the normal endpoint, to use these as proper SLIs, you would need to increase the scraping interval for **all** metrics on that endpoint and stuff like apiserver_request_total (i.e. high cardinality metrics) makes this cost-prohibitive. 

Your second thought has already been considered, it is one of the two metrics we are introducing. However, there are problems with counters, which is that there are situations where the counts are not super predictable and can really skew the ratios. This is why we supplement the counter with a gauge which can be scraped at exact intervals and computed against a bit more simply.

Reply:
We discussed this offline. I misunderstood @lavalamp. He was suggesting running a background goroutine which would periodically record the metrics at regular intervals. I had actually considered this approach but dismissed it because it was too invasive, since this needs to be applied to every single Kubernetes component.